### DISCUSSION TITLE
Trouble setting up a GraphQL proxy

### DISCUSSION THREAD

**Discussion by jonaird** [geldata/gel#3174] [Q&A]
Hi,

I'm having trouble setting up a proxy server for the GraphQL endpoint. I'm using Nodejs and Express and when I try I get "Unknown path" as a response. It doesn't seem to be coming from the express server as far as I can tell. Could this be coming from the EdgeDB instance?

**Comment by jaclarke** [comment]
EdgeDB does return a 404 response with the message 'Unknown path' for http requests with a path it doesn't know.

You may not have the graphql extension enabled for your db. If you run `describe schema as sdl;` in the CLI, the response should contain this line: `using extension graphql version '1.0'`. If not, you need to add `using extension graphql;` to your schema file and run the migration workflow (if you prefer using ddl, then running `create extension graphql;` in the CLI also works).

Then you can try `http://localhost:<instance-port>/db/<database-name>/graphql/explore` in the browser to test it out. (`database-name` by default is 'edgedb', and `instance-port` can be found by running `edgedb instance list`).

**Reply by jonaird** [reply]
Yeah the graphql endpoint is set up and queries work when hitting the endpoint directly but they don't work through the proxy for some reason. I'm using an Express plugin for the proxy. Maybe it's changing something about the request?

**Reply by jaclarke** [reply]
Do you have a simple repo that reproduces the issue you can share?

**Reply by jonaird** [reply]
It's only a few lines so I'll just paste it here:
```
const express = require('express');
const { createProxyMiddleware } = require('http-proxy-middleware');
const app = express();

const portNumber = 'edgedb port number';
const endpoint  = 'http://127.0.0.1:'+portNumber+'/db/edgedb/graphql';
app.use('/graphql',createProxyMiddleware({target:endpoint,changeOrigin:true}));
app.listen(3000);
```

**Reply by jonaird** [reply]
I changed the proxy endpoint to '/' rather Thant '/graphql' and it seems to have fixed the issue.

**Reply by jonaird** [reply]
To be clear, it's still an issue if you want to serve from any other endpoint but serving from '/' works for now. Here's the repo if needed:
https://github.com/jonaird/edgedb-proxy-bug

**Reply by jaclarke** [reply]
As far as I understand from the http-proxy-middleware docs, this is the expected behaviour. By default the `prependPath` option is true, so setting the proxy path to `/graphql` means the request is proxied to `/db/edgedb/graphql/graphql` which is an invalid endpoint. To have a proxy path other than `/`, I think you would need to use the `pathRewrite` option:
```
const endpoint = "http://127.0.0.1:" + portNumber + "/db/edgedb/graphql";
app.use(
  "/some/custom/endpoint",
  createProxyMiddleware({
    target: endpoint,
    pathRewrite: { "^/some/custom/endpoint": "" },
  })
);
```
